pkg/server/filters/filters.go

```diff
@@ -78,6 +86,49 @@ func WithAuditEventClusterAnnotation(handler http.Handler) http.HandlerFunc {
 	})
 }
 
+// WithBlockInactiveLogicalClusters ensures that any requests to logical
+// clusters marked inactive are rejected.
+func WithBlockInactiveLogicalClusters(handler http.Handler, kcpClusterClient informersv1alpha1.LogicalClusterClusterInformer) http.HandlerFunc {
+	allowedPathPrefixes := []string{
+		"/openapi",
+		"/apis/core.kcp.io/v1alpha1/logicalclusters",
+	}
+
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		_, newURL, _, err := ClusterPathFromAndStrip(req)
+		if err != nil {
+			responsewriters.InternalError(w, req, err)
+			return
+		}
+
+		isException := false
+		for _, prefix := range allowedPathPrefixes {
+			if strings.HasPrefix(newURL.String(), prefix) {
+				isException = true
+			}
+		}
+
+		cluster := request.ClusterFrom(req.Context())
+		if cluster != nil && !cluster.Name.Empty() && !isException {
+			logicalCluster, err := kcpClusterClient.Cluster(cluster.Name).Lister().Get(corev1alpha1.LogicalClusterName)
+			if err == nil {
+				if ann, ok := logicalCluster.ObjectMeta.Annotations[inactiveAnnotation]; ok && ann == "true" {
+					responsewriters.ErrorNegotiated(
+						apierrors.NewBadRequest("logical cluster is marked inactive"),
```

Comment:
400 is explicitly a client error. Looking through the list

403 Forbidden
423 Locked
503 Service Unavailable

What do we return when accessing a logical cluster without the `LogicalCluster` object? 403 I assume? Maybe that would be the right response here too.